compiler/hie-core/src/Development/IDE/Core/Debouncer.hs

```diff
@@ -0,0 +1,67 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Development.IDE.Core.Debouncer
+    ( MicroSeconds(..)
+    , Debouncer
+    , newDebouncer
+    , withDebouncer
+    , registerEvent
+    , killEvents
+    , waitForEvents
+    ) where
+
+import Control.Concurrent
+import Control.Concurrent.Extra
+import Control.Concurrent.Async
+import Control.Exception
+import Control.Monad.Extra
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+
+-- | A debouncer can be used to avoid triggering many events
+-- (e.g. diagnostics) for the same key (e.g. the same file)
+-- within a short timeframe. This is accomplished
+-- by delaying each event for a given time. If another event
+-- is registered for the same key within that timeframe,
+-- only the new event will fire.
+newtype Debouncer k = Debouncer (Var (Map k (Async ())))
+
+newtype MicroSeconds = MicroSeconds Int
```

Comment:
My preference is just to always use System.Time.Extra.Seconds for all seconds, since people don't think in microseconds very naturally and always get it wrong.